src/app/reporting/Read-DataModel.cpp

```diff
@@ -0,0 +1,97 @@
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include <app/reporting/Read-DataModel.h>
+
+#include <app/AppConfig.h>
+#include <app/data-model-interface/DataModel.h>
+#include <app/util/MatterCallbacks.h>
+
+namespace chip {
+namespace app {
+namespace reporting {
+namespace DataModelImpl {
+namespace {
+
+bool IsOutOfSpaceError(CHIP_ERROR err)
+{
+    return (err == CHIP_ERROR_NO_MEMORY || err == CHIP_ERROR_BUFFER_TOO_SMALL);
+}
+
+} // namespace
+
+CHIP_ERROR RetrieveClusterData(InteractionModel::DataModel * dataModel, const Access::SubjectDescriptor & subjectDescriptor,
+                               bool isFabricFiltered, AttributeReportIBs::Builder & reportBuilder,
+                               const ConcreteReadAttributePath & path, AttributeEncodeState * encoderState)
+{
+#if !CHIP_CONFIG_USE_EMBER_DATA_MODEL
+    ChipLogDetail(DataManagement, "<RE:Run> Cluster %" PRIx32 ", Attribute %" PRIx32 " is dirty", path.mClusterId,
+                  path.mAttributeId);
+    DataModelCallbacks::GetInstance()->AttributeOperation(DataModelCallbacks::OperationType::Read,
+                                                          DataModelCallbacks::OperationOrder::Pre, path);
+#endif // !CHIP_CONFIG_USE_EMBER_DATA_MODEL
+
+    InteractionModel::ReadAttributeRequest readRequest;
+
+    if (isFabricFiltered)
+    {
+        readRequest.readFlags.Set(InteractionModel::ReadFlags::kFabricFiltered);
+    }
+    readRequest.subjectDescriptor = subjectDescriptor;
+    readRequest.path              = path;
+
+    DataVersion version = 0;
+    if (std::optional<InteractionModel::ClusterInfo> clusterInfo = dataModel->GetClusterInfo(path); clusterInfo.has_value())
+    {
+        version = clusterInfo->dataVersion;
+    }
+    else
+    {
+        ChipLogError(DataManagement, "Read request on unknown cluster - no data version available");
+    }
+
+    TLV::TLVWriter checkpoint;
+    reportBuilder.Checkpoint(checkpoint);
+
+    AttributeValueEncoder attributeValueEncoder(reportBuilder, subjectDescriptor, path, version, isFabricFiltered, encoderState);
+    CHIP_ERROR err = dataModel->ReadAttribute(readRequest, attributeValueEncoder);
+
+    if (err == CHIP_NO_ERROR)
+    {
+#if !CHIP_CONFIG_USE_EMBER_DATA_MODEL
+        DataModelCallbacks::GetInstance()->AttributeOperation(DataModelCallbacks::OperationType::Read,
+                                                              DataModelCallbacks::OperationOrder::Post, path);
+#endif // !CHIP_CONFIG_USE_EMBER_DATA_MODEL
+        return CHIP_NO_ERROR;
+    }
+
+    if (IsOutOfSpaceError(err) || err == CHIP_IM_GLOBAL_STATUS(InvalidDataType))
+    {
+        // Only update state for retryable errors. This matches logic in legacy ember compatibility layer
+        if (encoderState != nullptr)
+        {
+            *encoderState = attributeValueEncoder.GetState();
```

Comment:
This does not make sense to me.  The "ember" code sets the state on all errors from AttributeAccessInterface (and never in other cases), right?  Why are we only doing it for some errors here?

Reply:
I did this for unit testing. Unfortunately interface-wise it makes no sense for the data provider to have information on "this is AAI or not".

I updated these to work and pass tests, will have to revert and add some descriptions. In my mind the correct thing to do would be to provide some guarantees/description like:
   - encoder state always forwarded (even if not useful)
   - encoder state is only for out of space, so it should be forwarded for out of space errors only
   
The logic of `encoder state forwarded if AAI is used, otherwise if ember is used it is not` ties our interface to implementation details (i.e. ember) and I believe that is not ok.

Reply:
No, my question is: why are we doing _any_ restriction on the error here?  Just pass through all errors as-is.

Reply:
Unit tests fail if I do that (because ember currently does not do it). Will run a test run to check which.

Reply:
> because ember currently does not do it

It sure does as far as I can tell.

Reply:
Er, note that this is about resetting the encoder state on all errors, to be clear.

Reply:
Apparently this used to fail, maybe when I updated on both failure and success or before I fixed some code paths. Updated to always set the state on error.